File: java/tutorials/tour-of-workflows-java/src/main/java/com/example/workflows/workflows/SignupWorkflow.java
package com.example.workflows.workflows;

import static com.example.workflows.utils.Utils.*;

import com.example.workflows.types.User;
import dev.restate.sdk.WorkflowContext;
import dev.restate.sdk.annotation.Workflow;

@Workflow
public class SignupWorkflow {

  @Workflow
  public boolean run(WorkflowContext ctx, User user) {
    String userId = ctx.key(); // workflow ID = user ID

    // Write to database
    boolean success = ctx.run("create", Boolean.class, () -> createUser(userId, user));
    if (!success) {
      return false;
    }

    // Call APIs
    ctx.run("activate", () -> activateUser(userId));
    ctx.run("welcome", () -> sendWelcomeEmail(user));

    return true;
  }
}

File: python/tutorials/tour-of-workflows-python/app/workflows/signup_workflow.py
import restate
from restate import WorkflowContext
from app.utils import User, create_user, activate_user, send_welcome_email

signup_workflow = restate.Workflow("SignupWorkflow")


@signup_workflow.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    user_id = ctx.key()  # workflow ID = user ID

    # Write to database
    success = await ctx.run_typed("create", create_user, user_id=user_id, user=user)
    if not success:
        return False

    # Call APIs
    await ctx.run_typed("activate", activate_user, user_id=user_id)
    await ctx.run_typed("welcome", send_welcome_email, user=user)
    return True
